src/createSlice.ts

```diff
@@ -84,84 +94,97 @@ export interface CreateSliceOptions<
     | ((builder: ActionReducerMapBuilder<NoInfer<State>>) => void)
 }
 
-type PayloadActions<Types extends keyof any = string> = Record<
-  Types,
-  PayloadAction
->
-
-type CaseReducerWithPrepare<State, Action extends PayloadAction> = {
+/**
+ * A CaseReducer with a `prepare` method.
+ *
+ * @public
+ */
+export type CaseReducerWithPrepare<State, Action extends PayloadAction> = {
   reducer: CaseReducer<State, Action>
   prepare: PrepareAction<Action['payload']>
 }
 
-type SliceCaseReducerDefinitions<State, PA extends PayloadActions> = {
-  [ActionType in keyof PA]:
-    | CaseReducer<State, PA[ActionType]>
-    | CaseReducerWithPrepare<State, PA[ActionType]>
-}
+/**
+ * The type describing a slice's `reducers` option.
+ * Also checks itself, so it has to be passed "itself" as it's second option.
```

Comment:
wat :)

Reply:
Yeah, I wouldn't have found out about that one myself as it is completely counter-intuitive. But a generic can `extend` (which essentially just means "match this shape") something that *depends on the generic being matched*.

So you can write
```ts
function something<X extends Restrict<X>>(x: X) {}
```

and then `Restrict` can be a type that does some crazy restrictions on the original input object while referencing it. 

This is something I learned from @jcalz over on StackOverflow: https://stackoverflow.com/questions/59200095/is-there-a-simpler-way-of-associating-a-key-to-a-value-in-a-mapped-obect?noredirect=1#comment104621839_59200095
Thank you again Joe!

Reply:
Still waiting for that blog post / talk on "all the bizarre stuff I learned working on RTK's TS types" :)

Reply:
Yup, got something in the making :) Meetup somewhere mid-february, maybe I'll also submit it to a few conferences. Also, some ideas to experiment with twitter-content :)